api/v1alpha1/azurevalidator_types.go

```diff
@@ -84,42 +85,80 @@ type CommunityGallery struct {
+	Scope string `json:"scope" yaml:"scope"`
+	// Role is the role data.
+	Role Role `json:"role" yaml:"role"`
+}
 
-// PermissionSet is part of an RBAC rule and verifies that a security principal has the specified
-// permissions (via role assignments) at the specified scope. Scope can be either subscription,
-// resource group, or resource.
-type PermissionSet struct {
-	// Actions is a list of actions that the role must be able to perform. Must not contain any
-	// wildcards. If not specified, the role is assumed to already be able to perform all required
-	// actions.
-	//+kubebuilder:validation:MaxItems=1000
-	//+kubebuilder:validation:XValidation:message="Actions cannot have wildcards.",rule="self.all(item, !item.contains('*'))"
+// Role is role data in a role assignment. Is it a subset of a role definition.
+type Role struct {
+	// Name is the role name property of the role definition.
+	// +kubebuilder:validation:MaxLength=200
+	Name string `json:"name" yaml:"name"`
+	// Type is the role type property of the role definition. Must be "BuiltInRole" or "Custom".
+	// Required to disambiguate built in roles and custom roles with the same name.
+	// +kubebuilder:validation:Enum=BuiltInRole;CustomRole
+	Type string `json:"type" yaml:"type"`
+	// Permission is the permissions data of the role definition.
+	Permission Permission `json:"permissions" yaml:"permissions"`
+}
+
+// Permission is the permission data in a role definition.
+type Permission struct {
+	// Actions is the "actions" of the role definition.
 	Actions []ActionStr `json:"actions,omitempty" yaml:"actions,omitempty"`
-	// DataActions is a list of data actions that the role must be able to perform. Must not
-	// contain any wildcards. If not provided, the role is assumed to already be able to perform
-	// all required data actions.
-	//+kubebuilder:validation:MaxItems=1000
-	//+kubebuilder:validation:XValidation:message="DataActions cannot have wildcards.",rule="self.all(item, !item.contains('*'))"
+	// DataActions is the "dataActions" of the role definition.
 	DataActions []ActionStr `json:"dataActions,omitempty" yaml:"dataActions,omitempty"`
-	// Scope is the minimum scope of the role. Role assignments found at higher level scopes will
-	// satisfy this. For example, a role assignment found with subscription scope will satisfy a
-	// permission set where the role scope specified is a resource group within that subscription.
-	Scope string `json:"scope" yaml:"scope"`
+	// NotActions is the "notActions" of the role definition.
+	NotActions []ActionStr `json:"notActions,omitempty" yaml:"notActions,omitempty"`
+	// NotDataActions is the "notDataActions" of the role definition.
+	NotDataActions []ActionStr `json:"notDataActions,omitempty" yaml:"notDataActions,omitempty"`
+}
+
+// Equal compares a Permission (from the spec) to an armauthorization.Permission (from the Azure API
+// response).
+func (p Permission) Equal(other armauthorization.Permission) bool {
```

Comment:
I guess you can't use the [Azure structs](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/authorization/armauthorization/models.go) directly because they lack YAML tags?

Reply:
Also, I think this needs to be reworked to use maps instead of assuming that the permissions will match up 100% by index between the two slices. We don't need other and p to match EXACTLY, right? It should be acceptable for other to be a superset of p.

Reply:
> I guess you can't use the [Azure structs](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/authorization/armauthorization/models.go) directly because they lack YAML tags?

That's right. Also, they lack `omitempty` which means plugin users can't omit them. They might want to, for example, only provide `actions` and omit the rest.

My thinking was that when it comes to struct re-use, we can't re-use the struct in the spec because it ends up mixing too many concerns together. But it makes sense to re-use their struct in validatorctl because it makes it easy for users to copy things in from Azure docs (or Spectro's) that are meant to be parsed using their types. Then, validatorctl would create the correct rules from that.

Reply:
> reworked to use maps instead of assuming that the permissions will match up 100% by index between the two slices. We don't need other and p to match EXACTLY, right? It should be acceptable for other to be a superset of p.

I agree. We want this to be strict with respect to making sure the role does what the user wants it to do, but we can allow users to create roles to be validated where the order of actions doesn't line up. It would all work the same. I'll change it.

Reply:
Actually I just realized you were talking about two separate things in that second point. Each needs to be addressed.

> reworked to use maps instead of assuming that the permissions will match up 100% by index between the two slices. We don't need other and p to match EXACTLY, right?

Correct. If we're validating that a role provides a particular set of permissions, the order they're specified in doesn't matter. So index doesn't matter.

> It should be acceptable for other to be a superset of p.

Can you confirm what you mean by this? My understanding was that the new rule should be stricter in that it would only pass validation if the role permissions matched exactly what was defined.

The rule that this is replacing worked like that, where validation passed as long as all roles assigned covered at least the permissions specified. And I concluded back when I made that rule that taking into account wildcards, "not actions", and "not data actions" was too complex for validation that worked like that. However, I could take another look at that.